mne/rank.py

```diff
@@ -211,24 +226,165 @@ def _get_rank_sss(inst):
+    # Using pca=True in compute_whitener might help.
+    return sum(compute_rank(inst, scalings, rank, info).values())
+
+
+def compute_rank(inst, scalings=None, rank=None, info=None):
+    """Compute the rank of data or noise covariance.
+
+    This function will normalize the rows of the data (typically
+    channels or vertices) such that non-zero singular values
+    should be close to one.
+
+    Parameters
+    ----------
+    inst : instance of Raw, Epochs, or Covariance
+        Raw measurements to compute the rank from or the covariance.
+    scalings : dict | None (default None)
+        Defaults to ``dict(mag=1e15, grad=1e13, eeg=1e6)``.
+        These defaults will scale different channel types
+        to comparable values.
+    rank : int | None | dict | 'info' | 'full'
+        This controls the rank computation that can be read from the
+        measurement info or estimated from the data. See ``Notes`` for
+        details. The default is ``'info'``.
+    info : instance of Info | None
+        The measurement info used to compute the covariance. It is
+        only necessary if inst is a Covariance object.
+
+    Returns
+    -------
+    rank : dict
+        Estimated rank of the data for each channel type.
+        To get the total rank, you can use ``sum(rank.values())``.
+
+    Notes
+    -----
+    The ``rank`` parameter can be:
+
+    :class:`python:int`
+        This value is used as the MEG rank. For other channel types,
+        rank is taken from ``info``.
+    :data:`python:None`
+        the rank will be automatically estimated
+        from the data after proper scaling of the different
+        channel types.
+    ``'info'`` (default)
+        Rank is inferred from `info`. If data have been processed
+        with Maxwell filtering, the Maxwell filtering header is used.
+        Otherwise, the channel counts themselves are used.
+        In both cases, the number of projectors is accounted for.
+    ``'full'``
+        Rank is assumed to be full, i.e. equal to the
+        number of good channels. If a `Covariance`
+        is passed, this can make sense if it has been (possibly
+        improperly) regularized without taking into account the true
+        data rank.
+
+    .. versionadded:: 0.18
+    """
+    from .io.base import BaseRaw
+    from .epochs import BaseEpochs
+    from . import Covariance
+
+    rank = _check_rank(rank)
+
+    if isinstance(inst, Covariance):
+        if info is None:
+            raise ValueError('info cannot be None if inst is a Covariance.')
+        inst = pick_channels_cov(
+            inst, set(inst['names']) & set(info['ch_names']))
+        info = pick_info(info, [info['ch_names'].index(name)
+                                for name in inst['names']])
     else:
-        raise ValueError('There is no `max_info` here. Sorry.')
-    return _get_sss_rank(max_info)
+        info = inst.info
+
+    if isinstance(rank, str):  # string
```

Comment:
string -> string : can be 'info' or 'full'

to make it explicit what it covers